### PR TITLE
Hassh feature 2698 v7

### PR DESCRIPTION
**Pull request by MalakhatkoVadym**
Link to redmine ticket:
https://redmine.openinfosecfoundation.org/issues/2698

Describe changes:
- Adds hassh and hasshServer for ssh fingerprinting to suricata.
- Adds lua logging to get hassh fingerprints.

Modifies https://github.com/OISF/suricata/pull/5091 by 
1. Adding limit for reassembled kex buffer;
2. Changed commit for `SSHTxLogCondition`;
3. Added `SSHTxLogCondition` for lua;
4. Small fixes.

### Testing

1. Enable hassh in suricata.yaml in app-layer.protocols.ssh.hassh:

```
app-layer:
  protocols:
    ssh:
      enabled: yes
      hassh: yes
```

2. Check your hassh parameters (hassh, server hassh, hassh string, hassh server string) (connect with ssh to tested host) using tool https://github.com/salesforce/hassh/tree/master/python 
or using suricata's lua scripting:
change suricata.yaml:
```
outputs:
  - lua:
      enabled: yes
      scripts-dir: /etc/suricata/lua-output/
      scripts:
         - hassh.lua
```
hassh.lua:

```
function init (args)
    local needs = {}
    needs["protocol"] = "ssh"
    return needs
end

function setup (args)
    filename = "/tmp/hassh.log"
    file = assert(io.open(filename, "a"))
end

function log(args)
    hassh = HasshGet()
    file:write (hassh .. " -hassh \n")

    hasshServer = HasshServerGet()
    file:write (hasshServer .. " -hasshServer \n")

    hasshString = HasshGetString()
    file:write (hasshString .. " -hasshString \n")

    hasshServerString = HasshServerGetString()
    file:write (hasshServerString .. " -hasshServerString \n")

    file:flush()
end

function deinit (args)
    file:close(file)
end

return 0
```

3. Add rules for hassh detection:
```
alert ssh any any -> any any (msg:"match SSH hash"; ssh.hassh; content:"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"; sid:1000010;)
alert ssh any any -> any any (msg:"match SSH hash-server"; ssh.hassh.server; content:"xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx"; sid:1000020;)
alert ssh any any -> any any (msg:"match SSH hash-string"; ssh.hassh.string; content:"none,zlib@openssh.com,zlib"; sid:1000030;)
alert ssh any any -> any any (msg:"match SSH hash-server-string"; ssh.hassh.server.string; content:"umac-64-etm@openssh.com,umac-128-etm@openssh.com,"; sid:1000040;)
```
4. Check alerts.

### CODE REVIEW

File: rust/src/ssh/ssh.rs [hunk 1]
```diff
@@ -179,8 +212,24 @@ impl SSHState {
...
+                                parser::MessageCode::SshMsgKexinit if HASSH_ENABLED.load(Ordering::Relaxed) => {
+                                    // check if buffer is bigger than maximum reassembled packet size
+                                    if hdr.record_left < SSH_MAX_REASSEMBLED_RECORD_LEN as u32 {
+                                        // saving type of incomplete kex message
+                                        hdr.record_left_msg = parser::MessageCode::SshMsgKexinit;
+                                        return AppLayerResult::incomplete(SSH_RECORD_HEADER_LEN as u32, hdr.record_left as u32);
+                                    }
+                                    else {
+                                        SCLogDebug!("SSH buffer is bigger than maximum reassembled packet size");
+                                        self.set_event(SSHEvent::InvalidRecord);
```
Comment (catenacyber) [inline rust/src/ssh/ssh.rs:228]: Can you set a new type of event here ?

File: rust/src/ssh/ssh.rs [hunk 2]
```diff
@@ -179,8 +212,24 @@ impl SSHState {
...
+                                        return AppLayerResult::err();
```
Comment (catenacyber) [inline rust/src/ssh/ssh.rs:229]: Why return error here ?
We can continue parsing by skipping this big record

Comment (MalakhatkoVadym) [reply rust/src/ssh/ssh.rs:229]: Should it return just `return AppLayerResult::ok()` ?

File: rust/src/ssh/ssh.rs (hunk 1 above)
Comment (MalakhatkoVadym) [reply rust/src/ssh/ssh.rs:228]: Yes. 
Can you, please, suggest a name for this event (There is an event for too long banner `SSHEvent::LongBanner`)?
`LongKex`?
`LongKexRecord`?
`KexRecordOverflow`?
just `RecordOverflow`?

File: rust/src/ssh/ssh.rs (hunk 2 above)
Comment (catenacyber) [reply rust/src/ssh/ssh.rs:229]: There should not be any `return` at all here, and let the function continue to the `return AppLayerResult::ok();` present a few lines below

File: rust/src/ssh/ssh.rs (hunk 1 above)
Comment (catenacyber) [reply rust/src/ssh/ssh.rs:228]: LongKexRecord looks good to me

Comment (MalakhatkoVadym) [reply rust/src/ssh/ssh.rs:228]: Thanks

**Comment by MalakhatkoVadym** [comment]
https://github.com/OISF/suricata/pull/5105

**Pull request closed without merging**